File: src/main/java/dev/todaka/robustredis/protocol/CommandOutput.java
package dev.todaka.robustredis.protocol;

import java.util.concurrent.CompletableFuture;

public abstract class CommandOutput<R> {
    protected CompletableFuture<R> future = new CompletableFuture<>();

    public CompletableFuture<R> getCompletableFuture() {
        return future;
    }
    
    public void setResult(String result) {
        future.completeExceptionally(new UnsupportedOperationException(
                "setResult(String) is not supported for this command"));
    }

    public void setResult(long result) {
        future.completeExceptionally(new UnsupportedOperationException(
                "setResult(long) is not supported for this command"));
    }

    public void setError(String error) {
        future.completeExceptionally(new RuntimeException(error));
    }
}

File: src/main/kotlin/dev/todaka/robustredis/CommandOutput.kt
package dev.todaka.robustredis

import java.util.concurrent.CompletableFuture

abstract class CommandOutput<R> {
    val completableFuture = CompletableFuture<R>()

    open fun setResult(result: String) {
        completableFuture.completeExceptionally(
            UnsupportedOperationException(
                "setResult(String) is not supported for this command"
            )
        )
    }

    open fun setResult(result: Long) {
        completableFuture.completeExceptionally(
            UnsupportedOperationException(
                "setResult(long) is not supported for this command"
            )
        )
    }

    open fun setError(error: String) {
        completableFuture.completeExceptionally(RuntimeException(error))
    }
}

class StringCommandOutput : CommandOutput<String>() {
    override fun setResult(result: String) {
        this.completableFuture.complete(result)
    }
}


class LongCommandOutput : CommandOutput<Long>() {
    override fun setResult(result: Long) {
        this.completableFuture.complete(result)
    }
}
